### ISSUE TITLE
训练sparse4d时，validate过程中，sparse.head出现了nan，导致报错

### ISSUE THREAD

**Issue by HuiAuto** [HorizonRobotics/Sparse4D#66]
您好！首先感谢您开源了sparse4d的代码供大家研究。

在我的训练环境中，cuda版本与python包的版本与requirment中严格相同，使用4090单GPU训练。为复现sparse4d的训练过程，在训练过程中我并未更改sparse4d_head及其它各部分的代码。

在训练过程中，可以通过输出发现loss_box_0等呈下降趋势，且不报错。
在val过程中，报错
![图片1](https://github.com/HorizonRobotics/Sparse4D/assets/138637360/f27c0c69-3c5c-4174-85e4-9fa379c078b1)
我使用了 has_nan = torch.any(torch.isnan(model_outs)) 在train和validate过程中逐步检测各模块输出，结果发现在训练的forward_train() 过程中没有出现任何nan值。在验证的simple_test()过程中出现nan值，nan值在

**model_outs = self.head(feature_maps, data)**

该语句后出现（feature_maps, data中无nan值，model_outs中出现nan值）。

**model_outs['classification']=[torch.tensor(1, 900, 10), None, None, None, None, torch.tensor(1, 900, 10)]，其中list中最后一位的tensor中元素全为nan。**

请问您在训练的时候有遇到这个问题吗？是怎么解决的呢？

**Comment by HuiAuto** [comment]
且在val过程中，对于6019个验证的sample，不是每一个都会出现nan，而是随机个数的sample出现nan。
偶尔第一次验证过程6019个验证的sample都不会出现nan，但第二次出现nan。
这个问题出现的原因是sparse4d_head中网络出现了问题吗？

**Comment by linxuewu** [comment]
训飞了吧，看看loss是不是正常在下降。

**Comment by HuiAuto** [comment]
感谢您的意见，您的意见是正确的，我发现在我的训练前期，loss在下降，grad_norm也与开源的训练结果相似。
但在单卡4090训练2300到2500个iter时，grad_norm出现异常变大。此后loss不再下降，grad_norm也变为nan。我重新下载了开源的代码，并重新从头开始尝试了训练，结果还是出现了这种情况。我现在有两个问题请教您一下：

1、请问这种情况是训飞了吗？
2、我在单卡训练时，只更改了config中的num_gpu=1和total_batch_size=6，其它的参数采用了和分布式8卡训练相同的参数。请问单卡训练是模型训飞的原因吗？如果是，大概应该向那个方向去调参呢？

.......

2024-07-04 21:58:28,546 - mmdet - INFO - Iter [2397/468800]	lr: 3.000e-04, eta: 3 days, 11:30:18, time: 0.635, data_time: 0.011, memory: 14353, loss_cls_0: 1.1075, loss_box_0: 2.1677, loss_cns_0: 0.5800, loss_yns_0: 0.1575, loss_cls_1: 1.2368, loss_box_1: 2.3636, loss_cns_1: 0.6192, loss_yns_1: 0.1659, loss_cls_2: 1.2174, loss_box_2: 2.2462, loss_cns_2: 0.6302, loss_yns_2: 0.1622, loss_cls_3: 1.1626, loss_box_3: 2.1932, loss_cns_3: 0.6284, loss_yns_3: 0.1661, loss_cls_4: 1.1289, loss_box_4: 2.1725, loss_cns_4: 0.6292, loss_yns_4: 0.1615, loss_cls_5: 1.1268, loss_box_5: 2.1571, loss_cns_5: 0.6319, loss_yns_5: 0.1623, loss_cls_dn_0: 0.2118, loss_box_dn_0: 0.9963, loss_cls_dn_1: 0.4788, loss_box_dn_1: 0.9511, loss_cls_dn_2: 0.3811, loss_box_dn_2: 0.8819, loss_cls_dn_3: 0.3021, loss_box_dn_3: 0.8533, loss_cls_dn_4: 0.2633, loss_box_dn_4: 0.8435, loss_cls_dn_5: 0.3303, loss_box_dn_5: 0.8381, loss_dense_depth: 0.9428, loss: 33.2489, grad_norm: 336.0426

2024-07-04 21:59:01,097 - mmdet - INFO - Iter [2448/468800]	lr: 3.000e-04, eta: 3 days, 11:28:44, time: 0.638, data_time: 0.011, memory: 14353, loss_cls_0: 1.0575, loss_box_0: 2.2432, loss_cns_0: 0.5760, loss_yns_0: 0.1648, loss_cls_1: 1.1661, loss_box_1: 2.5483, loss_cns_1: 0.5694, loss_yns_1: 0.1644, loss_cls_2: 1.1548, loss_box_2: 2.4796, loss_cns_2: 0.5775, loss_yns_2: 0.1677, loss_cls_3: 1.1357, loss_box_3: 2.3626, loss_cns_3: 0.6125, loss_yns_3: 0.1642, loss_cls_4: 1.0563, loss_box_4: 2.2616, loss_cns_4: 0.6196, loss_yns_4: 0.1540, loss_cls_5: 1.0235, loss_box_5: 2.2283, loss_cns_5: 0.6220, loss_yns_5: 0.1530, loss_cls_dn_0: 0.3238, loss_box_dn_0: 1.1059, loss_cls_dn_1: 0.4028, loss_box_dn_1: 1.2718, loss_cls_dn_2: 0.4085, loss_box_dn_2: 1.2229, loss_cls_dn_3: 0.4124, loss_box_dn_3: 1.1747, loss_cls_dn_4: 0.3781, loss_box_dn_4: 1.0867, loss_cls_dn_5: 0.3810, loss_box_dn_5: 1.0712, loss_dense_depth: 1.0882, loss: 35.5905, grad_norm: 618.2084

.......

2024-07-04 23:05:52,977 - mmdet - INFO - Iter [8721/468800]	lr: 2.997e-04, eta: 3 days, 9:54:29, time: 0.625, data_time: 0.012, memory: 14353, loss_cls_0: 0.8653, loss_box_0: 1.7395, loss_cns_0: 0.6306, loss_yns_0: 0.1607, loss_cls_1: 0.9439, loss_box_1: 1.6796, loss_cns_1: 0.6560, loss_yns_1: 0.1567, loss_cls_2: 0.9403, loss_box_2: 1.6390, loss_cns_2: 0.6582, loss_yns_2: 0.1577, loss_cls_3: 0.9489, loss_box_3: 1.6222, loss_cns_3: 0.6579, loss_yns_3: 0.1573, loss_cls_4: 0.9283, loss_box_4: 1.6126, loss_cns_4: 0.6564, loss_yns_4: 0.1568, loss_cls_5: 0.9253, loss_box_5: 1.6042, loss_cns_5: 0.6554, loss_yns_5: 0.1558, loss_cls_dn_0: 0.1263, loss_box_dn_0: 0.7506, loss_cls_dn_1: 0.1756, loss_box_dn_1: 0.7653, loss_cls_dn_2: 0.1433, loss_box_dn_2: 0.6957, loss_cls_dn_3: 0.1417, loss_box_dn_3: 0.6880, loss_cls_dn_4: 0.1368, loss_box_dn_4: 0.6822, loss_cls_dn_5: 0.1253, loss_box_dn_5: 0.6764, loss_dense_depth: 0.9325, loss: 26.3482, grad_norm: nan

2024-07-04 23:06:24,439 - mmdet - INFO - Iter [8772/468800]	lr: 2.997e-04, eta: 3 days, 9:52:52, time: 0.617, data_time: 0.012, memory: 14353, loss_cls_0: 0.9511, loss_box_0: 1.8558, loss_cns_0: 0.6217, loss_yns_0: 0.1584, loss_cls_1: 1.0059, loss_box_1: 1.8436, loss_cns_1: 0.6468, loss_yns_1: 0.1612, loss_cls_2: 1.0099, loss_box_2: 1.8050, loss_cns_2: 0.6491, loss_yns_2: 0.1602, loss_cls_3: 1.0228, loss_box_3: 1.7902, loss_cns_3: 0.6502, loss_yns_3: 0.1596, loss_cls_4: 1.0028, loss_box_4: 1.7819, loss_cns_4: 0.6486, loss_yns_4: 0.1586, loss_cls_5: 0.9984, loss_box_5: 1.7699, loss_cns_5: 0.6482, loss_yns_5: 0.1576, loss_cls_dn_0: 0.1051, loss_box_dn_0: 0.7487, loss_cls_dn_1: 0.1585, loss_box_dn_1: 0.7383, loss_cls_dn_2: 0.1299, loss_box_dn_2: 0.6784, loss_cls_dn_3: 0.1342, loss_box_dn_3: 0.6697, loss_cls_dn_4: 0.1270, loss_box_dn_4: 0.6643, loss_cls_dn_5: 0.1154, loss_box_dn_5: 0.6589, loss_dense_depth: 0.9146, loss: 27.5003, grad_norm: nan

........

**Comment by shryao1** [comment]
单卡训练好像很容易出现这个问题，后来我开了一下多卡，好像loss收敛就正常了，但是还是遇到了loss log variables are different across GPUs的bug

**Comment by linxuewu** [comment]
单卡，因为batch size小，所以需要减小学习率。你可以看看之前的issue。

loss log variables are different across GPUs，这应该某个gpu的某个batch 里面所有样本的gt数量都是0，导致没有dn_loss，gather的时候报错了。

**Comment by HuiAuto** [comment]
通过调整学习率，我解决了这个问题，十分感谢！

**Comment by xifen523** [comment]
> 通过调整学习率，我解决了这个问题，十分感谢！

可以分享分享一下具体如何调整吗？
其中有：
optimizer = dict(
    type="AdamW",
    lr=6e-4,
    weight_decay=0.001,
    paramwise_cfg=dict(
        custom_keys={
            "img_backbone": dict(lr_mult=0.5),
        }
    ),
)
optimizer_config = dict(grad_clip=dict(max_norm=25, norm_type=2))
lr_config = dict(
    policy="CosineAnnealing",
    warmup="linear",
    warmup_iters=500,
    warmup_ratio=1.0 / 3,
    min_lr_ratio=1e-3,
)

需要调整哪几个？

**Comment by HuiAuto** [comment]
optimizer = dict(
type="AdamW",
lr=6e-4,
weight_decay=0.001,
paramwise_cfg=dict(
custom_keys={
"img_backbone": dict(lr_mult=0.5),
}
),
)
这个里面的lr是学习率，batch_size缩小几倍就下调几倍，基本上就可以了

lr_config = dict(
policy="CosineAnnealing",
warmup="linear",
warmup_iters=500,
warmup_ratio=1.0 / 3,
min_lr_ratio=1e-3,
)
这个是控制学习率线性变化，使模型训练预热的，不用更改

**Comment by feimadada** [comment]
> 单卡训练好像很容易出现这个问题，后来我开了一下多卡，好像loss收敛就正常了，但是还是遇到了loss log variables are different across GPUs的bug

你好，这个bug你解决了吗
